File: backend/piclient.py
"""Drive a `pi --mode rpc` agent process over its JSONL stdin/stdout protocol.

One `PiSession` is one live call leg: a single agent process, either local (the
operator, on this box) or remote (a project agent, over ssh on the host that
holds the code). The switchboard writes `prompt` commands in and reads the event
stream back out, harvesting two things per turn:

  * the assistant's text, which becomes the spoken reply, and
  * *signals* — switchboard tool calls (`transfer_to_project`,
    `return_to_operator`) seen in the `tool_execution_start` stream.

Signals are how an agent moves the caller. The tools themselves (see
files/pi-extensions/) do nothing but acknowledge; the routing decision is made
here, by the process that actually owns the phone line.

Protocol notes that bite if ignored (docs/rpc.md in the pi package):
  * Strict JSONL, LF-delimited. Split on "\\n" only.
  * Event lines embed whole message objects and routinely exceed asyncio's
    default 64 KiB StreamReader limit, so the limit is raised explicitly below.
"""

from __future__ import annotations

import asyncio
import json
import logging
import shlex
from dataclasses import dataclass, field

log = logging.getLogger("switchboard.pi")

# Event lines carry full message history; 64 KiB (asyncio's default) overflows
# on any non-trivial turn and raises LimitOverrunError mid-conversation.
STREAM_LIMIT = 16 * 1024 * 1024

# Tools the switchboard interprets as signals rather than ordinary work.
TRANSFER_TOOL = "transfer_to_project"
RETURN_TOOL = "return_to_operator"
# Restarts the current leg on a different model. Like the routing tools it only
# raises a signal: an agent cannot re-launch itself, because the process it
# would have to replace is the one making the call.
SET_MODEL_TOOL = "set_model"
# Not a routing signal — the agent's extension already delivered this audio to
# the caller by the time we see it. It is watched so the switchboard knows the
# agent handled its own voice this turn and does not read the written reply out
# on top of what was already said.
SPEAK_TOOL = "speak"
SIGNAL_TOOLS = {TRANSFER_TOOL, RETURN_TOOL, SET_MODEL_TOOL, SPEAK_TOOL}

# Text fallback for runtimes that cannot load the pi extension (a project host
# without pi, a claude/codex session). Agents are told to emit this exact line;
# it means the same thing as calling RETURN_TOOL.
RETURN_SENTINEL = "[[SWITCHBOARD:RETURN]]"

# An agent whose model call failed still ends its turn cleanly: pi emits an
# assistant message with `stopReason: "error"`, empty content and zero tokens,
# then settles. Read only as "text plus signals" that is indistinguishable from
# an agent that had nothing to say, which is how an expired token on a project
# host turned into silence on the line instead of a handoff back to the
# operator.
ERROR_STOP_REASON = "error"

# `errorMessage` carries a full stack trace. This is spoken aloud, so keep the
# first line and cut it to something a person can listen to.
ERROR_DETAIL_CHARS = 160


@dataclass
class Signal:
    """A routing request an agent made during a turn."""

    name: str
    args: dict = field(default_factory=dict)


@dataclass
class Turn:
    """What one prompt produced: something to say, and possibly somewhere to go."""

    text: str
    signals: list[Signal] = field(default_factory=list)
    failed: bool = False
    # Why it failed, in a form fit to say out loud. Empty when nothing went
    # wrong, or when the failure left no explanation behind.
    error: str = ""

    @property
    def agent_spoke(self) -> bool:
        """Whether the agent said its piece itself during this turn."""
        return any(s.name == SPEAK_TOOL for s in self.signals)


class PiSessionError(RuntimeError):
    pass


def _spoken_error(detail: object) -> str:
    """Reduce a model-call error to a line that can be read out on a call.

    Always returns something: a failure the caller cannot be told the reason
    for is still a failure they need handing back to the operator for.
    """
    if not isinstance(detail, str) or not detail.strip():
        return "the model call failed"
    first = detail.strip().splitlines()[0].strip()
    # Everything from the first "; details=" or " url=" on is diagnostics for a
    # log, not for a person listening to it.
    for cut in ("; details=", " url=", "; stack="):
        head, sep, _ = first.partition(cut)
        if sep:
            first = head.strip()
    if len(first) > ERROR_DETAIL_CHARS:
        first = first[:ERROR_DETAIL_CHARS].rstrip() + "…"
    return first or "the model call failed"


class PiSession:
    """A single agent process, prompted one turn at a time."""

    def __init__(
        self,
        argv: list[str],
        *,
        label: str,
        cwd: str | None = None,
        env: dict[str, str] | None = None,
        turn_timeout: float = 600.0,
    ) -> None:
        self.argv = argv
        self.label = label
        self.cwd = cwd
        self.env = env
        self.turn_timeout = turn_timeout
        self._proc: asyncio.subprocess.Process | None = None
        self._stderr_tail: list[str] = []
        self._stderr_task: asyncio.Task | None = None
        # One turn at a time: the RPC protocol rejects a second `prompt` while
        # the agent is streaming unless a streamingBehavior is given, and a
        # phone call is strictly turn-taking anyway.
        self._lock = asyncio.Lock()

    # -- lifecycle ---------------------------------------------------------

    async def start(self) -> None:
        log.info("[%s] starting: %s", self.label, " ".join(self.argv))
        try:
            self._proc = await asyncio.create_subprocess_exec(
                *self.argv,
                stdin=asyncio.subprocess.PIPE,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
                cwd=self.cwd,
                env=self.env,
                limit=STREAM_LIMIT,
            )
        except OSError as exc:
            raise PiSessionError(f"could not start {self.argv[0]}: {exc}") from exc
        self._stderr_task = asyncio.create_task(self._drain_stderr())

    async def _drain_stderr(self) -> None:
        """Keep stderr flowing (a full pipe deadlocks the child) and keep a tail.

        The tail is what turns "the agent said nothing" into a diagnosable
        message — ssh auth failures and missing binaries only ever show up here.
        """
        assert self._proc is not None and self._proc.stderr is not None
        try:
            while True:
                line = await self._proc.stderr.readline()
                if not line:
                    return
                text = line.decode("utf-8", "replace").rstrip()
                if text:
                    log.debug("[%s] stderr: %s", self.label, text)
                    self._stderr_tail.append(text)
                    del self._stderr_tail[:-20]
        except asyncio.CancelledError:
            raise
        except Exception:  # noqa: BLE001
            log.exception("[%s] stderr drain failed", self.label)

    @property
    def alive(self) -> bool:
        return self._proc is not None and self._proc.returncode is None

    def stderr_tail(self, limit: int = 5) -> str:
        return " | ".join(self._stderr_tail[-limit:])

    async def close(self) -> None:
        """Hang up this leg. Best-effort and always safe to call twice."""
        proc, self._proc = self._proc, None
        if self._stderr_task is not None:
            self._stderr_task.cancel()
            self._stderr_task = None
        if proc is None or proc.returncode is not None:
            return
        log.info("[%s] closing", self.label)
        try:
            if proc.stdin is not None and not proc.stdin.is_closing():
                proc.stdin.close()
        except Exception:  # noqa: BLE001
            pass
        try:
            proc.terminate()
        except ProcessLookupError:
            return
        except Exception:  # noqa: BLE001
            pass
        try:
            await asyncio.wait_for(proc.wait(), timeout=10)
        except (asyncio.TimeoutError, Exception):  # noqa: BLE001
            try:
                proc.kill()
            except Exception:  # noqa: BLE001
                pass

    # -- one turn ----------------------------------------------------------

    async def prompt(self, message: str) -> Turn:
        """Send one user message and collect the reply plus any routing signals."""
        async with self._lock:
            if not self.alive:
                raise PiSessionError(
                    f"agent process is not running ({self.stderr_tail() or 'no output'})"
                )
            assert self._proc is not None
            assert self._proc.stdin is not None

            payload = json.dumps({"type": "prompt", "message": message}) + "\n"
            try:
                self._proc.stdin.write(payload.encode())
                await self._proc.stdin.drain()
            except (BrokenPipeError, ConnectionResetError) as exc:
                raise PiSessionError(f"agent process closed its input: {exc}") from exc

            try:
                return await asyncio.wait_for(self._collect(), timeout=self.turn_timeout)
            except asyncio.TimeoutError:
                log.warning("[%s] turn timed out after %ss", self.label, self.turn_timeout)
                return Turn(
                    text="That is taking longer than a phone call should. I have stopped waiting.",
                    failed=True,
                )

    async def _collect(self) -> Turn:
        """Read events until the agent settles, harvesting text and signals."""
        assert self._proc is not None and self._proc.stdout is not None
        stdout = self._proc.stdout

        chunks: list[str] = []
        signals: list[Signal] = []
        error = ""

        while True:
            try:
                raw = await stdout.readline()
            except (asyncio.LimitOverrunError, ValueError) as exc:
                # A single event exceeded STREAM_LIMIT. The line is unusable and
                # the stream is mid-record, so this leg is finished.
                log.error("[%s] oversized RPC event: %s", self.label, exc)
                return Turn(text="", signals=signals, failed=True)

            if not raw:
                log.warning("[%s] agent stream ended mid-turn", self.label)
                return Turn(text="".join(chunks), signals=signals, failed=True)

            line = raw.decode("utf-8", "replace").rstrip("\r\n")
            if not line:
                continue
            try:
                event = json.loads(line)
            except json.JSONDecodeError:
                log.debug("[%s] non-JSON line: %.200s", self.label, line)
                continue

            kind = event.get("type")

            if kind == "message_update":
                # Take completed text blocks rather than accumulating deltas —
                # `text_end` carries the whole block, so there is nothing to
                # reassemble and no risk of double-counting a retried stream.
                delta = event.get("assistantMessageEvent") or {}
                if delta.get("type") == "text_end":
                    content = delta.get("content")
                    if isinstance(content, str) and content.strip():
                        chunks.append(content.strip())

            elif kind == "message_end":
                # The only place a failed model call is reported. It is not an
                # `extension_error` and it does not break the stream: the turn
                # settles normally with an empty assistant message, so without
                # this the caller just hears nothing.
                message = event.get("message") or {}
                if (
                    message.get("role") == "assistant"
                    and message.get("stopReason") == ERROR_STOP_REASON
                ):
                    error = _spoken_error(message.get("errorMessage"))
                    log.error("[%s] model call failed: %s", self.label, message.get("errorMessage"))

            elif kind == "tool_execution_start":
                name = event.get("toolName")
                if name in SIGNAL_TOOLS:
                    args = event.get("args")
                    signals.append(Signal(name=name, args=args if isinstance(args, dict) else {}))
                    log.info("[%s] signal: %s %s", self.label, name, args)

            elif kind == "agent_settled":
                break

            elif kind == "extension_error":
                log.error("[%s] extension error: %s", self.label, event.get("error"))

        text = "\n".join(chunks).strip()

        # Sentinel fallback for runtimes without the switchboard extension.
        if RETURN_SENTINEL in text:
            text = text.replace(RETURN_SENTINEL, "").strip()
            if not any(s.name == RETURN_TOOL for s in signals):
                signals.append(Signal(name=RETURN_TOOL, args={"via": "sentinel"}))

        # A turn that errored but still produced something — a partial answer, a
        # tool call that landed before the failure — is worth keeping. It is the
        # error with nothing to show for it that has to be surfaced as a
        # failure, so the caller is put back with the operator.
        return Turn(text=text, signals=signals, failed=bool(error), error=error)


def list_models_argv(binary: str, ssh_host: str = "") -> list[str]:
    """Build the argv that prints a host's model catalog.

    Which host matters: a project agent runs where its code is, and that box has
    its own providers configured, so asking damocles what exists would answer
    for the wrong machine.
    """
    if not ssh_host:
        return [binary, "--list-models"]
    return [
        "ssh",
        "-T",
        "-o",
        "BatchMode=yes",
        "-o",
        "ConnectTimeout=10",
        ssh_host,
        f"{shlex.quote(binary)} --list-models",
    ]


def local_argv(
    binary: str,
    *,
    model: str | None,
    system_prompt_file: str | None,
    extension: str | None,
    extra_args: list[str] | None = None,
) -> list[str]:
    """Build the argv for an agent running on this box (the operator)."""
    argv = [binary, "--mode", "rpc"]
    if model:
        argv += ["--model", model]
    if system_prompt_file:
        argv += ["--system-prompt", _read_text(system_prompt_file)]
    if extension:
        argv += ["-e", extension]
    argv += extra_args or []
    return argv


def remote_argv(
    ssh_host: str,
    cwd: str,
    *,
    binary: str,
    model: str | None,
    extension: str | None,
    append_system_prompt: str | None,
    session_id: str | None = None,
    extra_args: list[str] | None = None,
    env: dict[str, str] | None = None,
) -> list[str]:
    """Build the argv for an agent running on the host that holds the project.

    `-T` because there is no terminal on either end, and BatchMode so a missing
    key fails fast and loudly on stderr instead of hanging on a password prompt.

    `env` is set inside the remote shell rather than passed through ssh, because
    sshd only accepts environment variables its config explicitly permits and
    silently drops the rest — which would leave the agent unable to speak with
    no indication why.
    """
    remote: list[str] = [binary, "--mode", "rpc"]
    if model:
        remote += ["--model", model]
    # A fixed id is what lets a leg be restarted on another model without losing
    # the conversation: the replacement process reopens the same session file.
    if session_id:
        remote += ["--session-id", session_id]
    if extension:
        remote += ["-e", extension]
    if append_system_prompt:
        remote += ["--append-system-prompt", append_system_prompt]
    remote += extra_args or []

    # `export NAME=value;` rather than the `NAME=value cmd` prefix form, because
    # the prefix form is not valid in front of `exec` — the shell would take the
    # assignment itself as the command to run.
    exports = "".join(
        f"export {name}={shlex.quote(value)}; " for name, value in sorted((env or {}).items())
    )
    # `set -e` so a bad cwd aborts instead of quietly starting the agent in the
    # ssh user's home directory — an agent in the wrong repo is worse than one
    # that never came up, because it looks like it worked.
    command = (
        f"set -e; cd {shlex.quote(cwd)}; {exports}"
        f"exec {' '.join(shlex.quote(a) for a in remote)}"
    )
    return [
        "ssh",
        "-T",
        "-o",
        "BatchMode=yes",
        "-o",
        "ConnectTimeout=10",
        ssh_host,
        command,
    ]


def _read_text(path: str) -> str:
    with open(path, encoding="utf-8") as fh:
        return fh.read()

File: src/pi_client.rs
use serde_json::{json, Map, Value};
use std::collections::HashMap;
use std::fmt;
use std::future::Future;
use std::path::Path;
use std::pin::Pin;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::{Arc, Mutex as StdMutex};
use tokio::io::{AsyncBufReadExt, AsyncWriteExt, BufReader};
use tokio::process::{Child, ChildStderr, ChildStdin, Command};
use tokio::sync::Mutex;
use tokio::time::{timeout, Duration};

pub const STREAM_LIMIT: usize = 16 * 1024 * 1024;
pub const TRANSFER_TOOL: &str = "transfer_to_project";
pub const RETURN_TOOL: &str = "return_to_operator";
pub const SET_MODEL_TOOL: &str = "set_model";
pub const SPEAK_TOOL: &str = "speak";
pub const RETURN_SENTINEL: &str = "[[SWITCHBOARD:RETURN]]";
const ERROR_STOP_REASON: &str = "error";
const ERROR_DETAIL_CHARS: usize = 160;
const ACTIVITY_DETAIL_CHARS: usize = 80;
const ACTIVITY_ARG_ORDER: [&str; 11] = [
    "path",
    "file_path",
    "filePath",
    "command",
    "pattern",
    "query",
    "url",
    "description",
    "symbol",
    "project",
    "text",
];

#[derive(Clone, Debug, PartialEq)]
pub struct Signal {
    pub name: String,
    pub args: Map<String, Value>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct Turn {
    pub text: String,
    pub signals: Vec<Signal>,
    pub failed: bool,
    pub error: String,
}
impl Turn {
    pub fn agent_spoke(&self) -> bool {
        self.signals.iter().any(|signal| signal.name == SPEAK_TOOL)
    }
}

#[derive(Debug)]
pub struct PiSessionError(pub String);
impl fmt::Display for PiSessionError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(&self.0)
    }
}
impl std::error::Error for PiSessionError {}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Activity {
    pub state: String,
    pub tool: String,
    pub detail: String,
    pub label: String,
}
pub type ActivityCallback =
    Arc<dyn Fn(Activity) -> Pin<Box<dyn Future<Output = ()> + Send>> + Send + Sync>;

struct SessionInner {
    child: Mutex<Option<Child>>,
    stdin: Mutex<Option<ChildStdin>>,
    stdout: Mutex<BufReader<tokio::process::ChildStdout>>,
    stderr_tail: Arc<StdMutex<Vec<String>>>,
    busy: AtomicBool,
    turn_lock: Mutex<()>,
    label: String,
    turn_timeout: Duration,
    on_activity: Option<ActivityCallback>,
    stderr_task: StdMutex<Option<tokio::task::JoinHandle<()>>>,
}

#[derive(Clone)]
pub struct PiSession {
    inner: Arc<SessionInner>,
    pub argv: Vec<String>,
    pub cwd: Option<String>,
}

impl PiSession {
    pub async fn start(
        argv: Vec<String>,
        label: impl Into<String>,
        cwd: Option<String>,
        env: Option<HashMap<String, String>>,
        turn_timeout: Duration,
        on_activity: Option<ActivityCallback>,
    ) -> Result<Self, PiSessionError> {
        let mut command = Command::new(
            argv.first()
                .ok_or_else(|| PiSessionError("no agent command was supplied".into()))?,
        );
        command
            .args(argv.iter().skip(1))
            .stdin(std::process::Stdio::piped())
            .stdout(std::process::Stdio::piped())
            .stderr(std::process::Stdio::piped())
            .kill_on_drop(true);
        if let Some(cwd) = &cwd {
            command.current_dir(cwd);
        }
        if let Some(env) = &env {
            command.envs(env);
        }
        let mut child = command
            .spawn()
            .map_err(|error| PiSessionError(format!("could not start {}: {error}", argv[0])))?;
        let stdin = child
            .stdin
            .take()
            .ok_or_else(|| PiSessionError("agent process has no stdin".into()))?;
        let stdout = child
            .stdout
            .take()
            .ok_or_else(|| PiSessionError("agent process has no stdout".into()))?;
        let stderr = child
            .stderr
            .take()
            .ok_or_else(|| PiSessionError("agent process has no stderr".into()))?;
        let stderr_tail = Arc::new(StdMutex::new(Vec::new()));
        let tail = Arc::clone(&stderr_tail);
        let label = label.into();
        let stderr_task = tokio::spawn(drain_stderr(stderr, tail));
        let inner = Arc::new(SessionInner {
            child: Mutex::new(Some(child)),
            stdin: Mutex::new(Some(stdin)),
            stdout: Mutex::new(BufReader::new(stdout)),
            stderr_tail,
            busy: AtomicBool::new(false),
            turn_lock: Mutex::new(()),
            label,
            turn_timeout,
            on_activity,
            stderr_task: StdMutex::new(Some(stderr_task)),
        });
        Ok(Self { inner, argv, cwd })
    }

    pub fn busy(&self) -> bool {
        self.inner.busy.load(Ordering::Acquire)
    }
    pub fn same_session(&self, other: &Self) -> bool {
        Arc::ptr_eq(&self.inner, &other.inner)
    }
    pub async fn alive(&self) -> bool {
        let mut child = self.inner.child.lock().await;
        child
            .as_mut()
            .is_some_and(|child| child.try_wait().ok().flatten().is_none())
    }
    pub fn stderr_tail(&self, limit: usize) -> String {
        self.inner
            .stderr_tail
            .lock()
            .map(|tail| {
                tail.iter()
                    .rev()
                    .take(limit)
                    .rev()
                    .cloned()
                    .collect::<Vec<_>>()
                    .join(" | ")
            })
            .unwrap_or_default()
    }

    pub async fn close(&self) {
        self.inner.busy.store(false, Ordering::Release);
        self.inner.stdin.lock().await.take();
        if let Some(mut child) = self.inner.child.lock().await.take() {
            let _ = child.kill().await;
            let _ = child.wait().await;
        }
        if let Ok(mut task) = self.inner.stderr_task.lock() {
            if let Some(task) = task.take() {
                task.abort();
            }
        }
    }

    pub async fn prompt(&self, message: &str) -> Result<Turn, PiSessionError> {
        let _turn = self.inner.turn_lock.lock().await;
        if !self.alive().await {
            return Err(PiSessionError(format!(
                "agent process is not running ({})",
                self.stderr_tail(5).as_str()
            )));
        }
        self.write(json!({"type":"prompt", "message":message}), false)
            .await?;
        self.inner.busy.store(true, Ordering::Release);
        let result = self.collect().await;
        self.inner.busy.store(false, Ordering::Release);
        result
    }

    pub async fn steer(&self, message: &str) -> Result<(), PiSessionError> {
        if !self.alive().await {
            return Err(PiSessionError(format!(
                "agent process is not running ({})",
                self.stderr_tail(5)
            )));
        }
        self.write(json!({"type":"steer", "message":message}), true)
            .await
    }

    async fn write(&self, command: Value, require_busy: bool) -> Result<(), PiSessionError> {
        let mut stdin = self.inner.stdin.lock().await;
        if require_busy && !self.busy() {
            return Err(PiSessionError("agent turn is no longer running".into()));
        }
        let stdin = stdin
            .as_mut()
            .ok_or_else(|| PiSessionError("agent process has no input to write to".into()))?;
        let mut payload =
            serde_json::to_vec(&command).map_err(|error| PiSessionError(error.to_string()))?;
        payload.push(b'\n');
        stdin
            .write_all(&payload)
            .await
            .map_err(|error| PiSessionError(format!("agent process closed its input: {error}")))?;
        stdin
            .flush()
            .await
            .map_err(|error| PiSessionError(format!("agent process closed its input: {error}")))
    }

    async fn collect(&self) -> Result<Turn, PiSessionError> {
        let mut chunks = Vec::new();
        let mut signals = Vec::new();
        let mut error = String::new();
        loop {
            let mut line = String::new();
            let read = {
                let mut stdout = self.inner.stdout.lock().await;
                timeout(self.inner.turn_timeout, stdout.read_line(&mut line)).await
            };
            let size = match read {
                Ok(Ok(size)) => size,
                Ok(Err(error)) => {
                    return Err(PiSessionError(format!(
                        "could not read agent output: {error}"
                    )))
                }
                Err(_) => {
                    self.close().await;
                    return Ok(Turn {
                        text: String::new(),
                        signals,
                        failed: true,
                        error: "the agent stopped responding".into(),
                    });
                }
            };
            if size == 0 {
                return Ok(Turn {
                    text: chunks.join("\n"),
                    signals,
                    failed: true,
                    error,
                });
            }
            if line.len() > STREAM_LIMIT {
                self.close().await;
                return Ok(Turn {
                    text: String::new(),
                    signals,
                    failed: true,
                    error: "the agent sent something too big to read".into(),
                });
            }
            let line = line.trim_end_matches(['\r', '\n']);
            if line.is_empty() {
                continue;
            }
            let event: Value = match serde_json::from_str(line) {
                Ok(event) => event,
                Err(_) => continue,
            };
            match event.get("type").and_then(Value::as_str) {
                Some("message_update") => {
                    if event
                        .pointer("/assistantMessageEvent/type")
                        .and_then(Value::as_str)
                        == Some("text_end")
                    {
                        if let Some(content) = event
                            .pointer("/assistantMessageEvent/content")
                            .and_then(Value::as_str)
                            .filter(|content| !content.trim().is_empty())
                        {
                            chunks.push(content.trim().to_owned());
                        }
                    }
                }
                Some("message_end") => {
                    let message = event.get("message").and_then(Value::as_object);
                    if message
                        .and_then(|message| message.get("role"))
                        .and_then(Value::as_str)
                        == Some("assistant")
                        && message
                            .and_then(|message| message.get("stopReason"))
                            .and_then(Value::as_str)
                            == Some(ERROR_STOP_REASON)
                    {
                        error =
                            spoken_error(message.and_then(|message| message.get("errorMessage")));
                    }
                }
                Some("tool_execution_start") => {
                    let name = event.get("toolName").and_then(Value::as_str).unwrap_or("");
                    if [TRANSFER_TOOL, RETURN_TOOL, SET_MODEL_TOOL, SPEAK_TOOL].contains(&name) {
                        signals.push(Signal {
                            name: name.into(),
                            args: event
                                .get("args")
                                .and_then(Value::as_object)
                                .cloned()
                                .unwrap_or_default(),
                        });
                    }
                    self.report_activity("start", name, activity_detail(event.get("args")))
                        .await;
                }
                Some("tool_execution_end") => {
                    self.report_activity(
                        "end",
                        event.get("toolName").and_then(Value::as_str).unwrap_or(""),
                        String::new(),
                    )
                    .await
                }
                Some("agent_settled") => break,
                _ => {}
            }
        }
        let mut text = chunks.join("\n").trim().to_owned();
        if text.contains(RETURN_SENTINEL) {
            text = text.replace(RETURN_SENTINEL, "").trim().to_owned();
            if !signals.iter().any(|signal| signal.name == RETURN_TOOL) {
                signals.push(Signal {
                    name: RETURN_TOOL.into(),
                    args: Map::from_iter([(String::from("via"), Value::String("sentinel".into()))]),
                });
            }
        }
        Ok(Turn {
            text,
            signals,
            failed: !error.is_empty(),
            error,
        })
    }

    async fn report_activity(&self, state: &str, tool: &str, detail: String) {
        if let Some(callback) = &self.inner.on_activity {
            callback(Activity {
                state: state.into(),
                tool: tool.into(),
                detail,
                label: self.inner.label.clone(),
            })
            .await;
        }
    }
}

async fn drain_stderr(stderr: ChildStderr, tail: Arc<StdMutex<Vec<String>>>) {
    let mut lines = BufReader::new(stderr).lines();
    while let Ok(Some(line)) = lines.next_line().await {
        if line.trim().is_empty() {
            continue;
        }
        if let Ok(mut tail) = tail.lock() {
            tail.push(line);
            if tail.len() > 20 {
                let remove = tail.len() - 20;
                tail.drain(..remove);
            }
        }
    }
}

fn activity_detail(args: Option<&Value>) -> String {
    let Some(args) = args.and_then(Value::as_object) else {
        return String::new();
    };
    for key in ACTIVITY_ARG_ORDER {
        if let Some(value) = args
            .get(key)
            .and_then(Value::as_str)
            .filter(|value| !value.trim().is_empty())
        {
            let value = value.split_whitespace().collect::<Vec<_>>().join(" ");
            if value.chars().count() > ACTIVITY_DETAIL_CHARS {
                let clipped = value
                    .chars()
                    .take(ACTIVITY_DETAIL_CHARS)
                    .collect::<String>();
                return format!("{}…", clipped.trim_end());
            }
            return value;
        }
    }
    String::new()
}

fn spoken_error(detail: Option<&Value>) -> String {
    let mut first = detail
        .and_then(Value::as_str)
        .unwrap_or("the model call failed")
        .trim()
        .split('\n')
        .next()
        .unwrap_or("")
        .to_owned();
    for cut in ["; details=", " url=", "; stack="] {
        if let Some(index) = first.find(cut) {
            first.truncate(index);
        }
    }
    if first.chars().count() > ERROR_DETAIL_CHARS {
        first = first.chars().take(ERROR_DETAIL_CHARS).collect();
        first.push('…');
    }
    if first.trim().is_empty() {
        "the model call failed".into()
    } else {
        first
    }
}

pub fn shell_quote(value: &str) -> String {
    format!("'{}'", value.replace('\'', "'\"'\"'"))
}
pub fn list_models_argv(binary: &str, ssh_host: &str) -> Vec<String> {
    if ssh_host.is_empty() {
        return vec![binary.into(), "--list-models".into()];
    }
    vec![
        "ssh".into(),
        "-T".into(),
        "-o".into(),
        "BatchMode=yes".into(),
        "-o".into(),
        "ConnectTimeout=10".into(),
        ssh_host.into(),
        format!("{} --list-models", shell_quote(binary)),
    ]
}
pub fn local_argv(
    binary: &str,
    model: Option<&str>,
    system_prompt_file: Option<&Path>,
    extension: Option<&str>,
    extra_args: &[String],
) -> Result<Vec<String>, PiSessionError> {
    let mut argv = vec![binary.into(), "--mode".into(), "rpc".into()];
    if let Some(model) = model {
        argv.extend(["--model".into(), model.into()]);
    }
    if let Some(path) = system_prompt_file {
        let prompt = std::fs::read_to_string(path).map_err(|error| {
            PiSessionError(format!(
                "could not read system prompt {}: {error}",
                path.display()
            ))
        })?;
        argv.extend(["--system-prompt".into(), prompt]);
    }
    if let Some(extension) = extension {
        argv.extend(["-e".into(), extension.into()]);
    }
    argv.extend(extra_args.iter().cloned());
    Ok(argv)
}
#[allow(clippy::too_many_arguments)]
pub fn remote_argv(
    host: &str,
    cwd: &str,
    binary: &str,
    model: Option<&str>,
    extension: Option<&str>,
    append_system_prompt: Option<&str>,
    session_id: Option<&str>,
    extra_args: &[String],
    env: &HashMap<String, String>,
) -> Vec<String> {
    let mut remote = vec![binary.into(), "--mode".into(), "rpc".into()];
    if let Some(model) = model {
        remote.extend(["--model".into(), model.into()]);
    }
    if let Some(session_id) = session_id {
        remote.extend(["--session-id".into(), session_id.into()]);
    }
    if let Some(extension) = extension {
        remote.extend(["-e".into(), extension.into()]);
    }
    if let Some(prompt) = append_system_prompt {
        remote.extend(["--append-system-prompt".into(), prompt.into()]);
    }
    remote.extend(extra_args.iter().cloned());
    let exports = env
        .iter()
        .map(|(name, value)| format!("export {name}={}; ", shell_quote(value)))
        .collect::<Vec<_>>()
        .join("");
    let command = format!(
        "set -e; cd {}; {}exec {}",
        shell_quote(cwd),
        exports,
        remote
            .iter()
            .map(|arg| shell_quote(arg))
            .collect::<Vec<_>>()
            .join(" ")
    );
    vec![
        "ssh".into(),
        "-T".into(),
        "-o".into(),
        "BatchMode=yes".into(),
        "-o".into(),
        "ConnectTimeout=10".into(),
        host.into(),
        command,
    ]
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn quotes_shell_values_and_builds_remote_commands() {
        assert_eq!(shell_quote("a b; rm -rf /"), "'a b; rm -rf /'");
        let args = remote_argv(
            "host",
            "/tmp/a b; rm -rf /",
            "pi",
            None,
            None,
            None,
            None,
            &[],
            &HashMap::new(),
        );
        assert!(args.last().unwrap().contains("cd '/tmp/a b; rm -rf /'"));
        assert!(args.last().unwrap().contains("exec 'pi' '--mode' 'rpc'"));
    }
    #[test]
    fn builds_local_rpc_argv() {
        let args = local_argv(
            "pi",
            Some("anthropic/opus"),
            None,
            Some("/tmp/ext.ts"),
            &["--no-session".into()],
        )
        .unwrap();
        assert_eq!(&args[..3], ["pi", "--mode", "rpc"]);
        assert!(args.contains(&"--no-session".into()));
    }
    #[test]
    fn parses_sentinel_and_activity_detail() {
        assert_eq!(
            activity_detail(Some(&json!({"command":"  ls   -la  "}))),
            "ls -la"
        );
        assert_eq!(
            activity_detail(Some(&json!({"command":"x".repeat(100)}))).len(),
            83
        );
        let unicode = "é".repeat(100);
        let detail = activity_detail(Some(&json!({"command": unicode})));
        assert!(detail.chars().count() <= ACTIVITY_DETAIL_CHARS + 1);
    }
    #[test]
    fn spoken_error_removes_diagnostics() {
        assert_eq!(
            spoken_error(Some(&json!(
                "OAuth failed url=https://example.test; details=secret"
            ))),
            "OAuth failed"
        );
        let unicode = spoken_error(Some(&json!("é".repeat(ERROR_DETAIL_CHARS + 10))));
        assert!(unicode.chars().count() <= ERROR_DETAIL_CHARS + 1);
        assert_eq!(spoken_error(None), "the model call failed");
    }

    #[tokio::test]
    async fn steer_writes_into_the_running_process() {
        let script = "read first; read second; printf '%s\\n' '{\"type\":\"agent_settled\"}'";
        let session = Arc::new(
            PiSession::start(
                vec!["sh".into(), "-c".into(), script.into()],
                "test",
                None,
                None,
                Duration::from_secs(1),
                None,
            )
            .await
            .unwrap(),
        );
        let running = Arc::clone(&session);
        let prompt = tokio::spawn(async move { running.prompt("hello").await.unwrap() });
        for _ in 0..10 {
            if session.busy() {
                break;
            }
            tokio::task::yield_now().await;
        }
        assert!(session.busy());
        session.steer("also check docs").await.unwrap();
        assert!(!prompt.await.unwrap().failed);
        session.close().await;
    }

    #[tokio::test]
    async fn process_prompt_collects_text_signal_and_sentinel() {
        let script = "read line; printf '%s\\n' '{\"type\":\"message_update\",\"assistantMessageEvent\":{\"type\":\"text_end\",\"content\":\"All done. [[SWITCHBOARD:RETURN]]\"}}' '{\"type\":\"agent_settled\"}'";
        let session = PiSession::start(
            vec!["sh".into(), "-c".into(), script.into()],
            "test",
            None,
            None,
            Duration::from_secs(1),
            None,
        )
        .await
        .unwrap();
        let turn = session.prompt("hello").await.unwrap();
        assert_eq!(turn.text, "All done.");
        assert_eq!(turn.signals[0].name, RETURN_TOOL);
        session.close().await;
    }
}
